src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -977,6 +979,7 @@ private void WriteMultiSegment(ReadOnlySpan<byte> source)
         {
             Span<byte> destination = _writingHeadMemory.Span;
 
+            Debug.Assert(_writingHead != null);
```

Comment:
Nit: this is really a precondition for the method: it'd be better if it were the first line